### PR TITLE
Recognise actors other than user in current_auth (resolves #160)

### LINKED ISSUE #160: current_auth API
Coaster received an auth framework in #154 and #155 with a proxy named `current_auth`. The proxy expects a login manager attached to the app that implements a `_load_user` method. However, as demonstrated in hasgeek/lastuser#229 (the first login manager to attempt a full switch to `current_auth`), the login manager has two distinct phases during a request:

1. Before request, when cookies are processed and a user is loaded.
2. After request, when a cookie is returned and activity logs are written.

Using `current_auth` in its current form replaces the first. It triggers a check for a user only when first required, so requests that don't require auth don't run the code. It also allows other `before_request` handlers to require a user without risk of being called before the user is loaded.

We now need a protocol for the second phase. Previously, we assumed the first phase runs on every request, so post-processing is required for every request. Now we need a flag somewhere to indicate `current_auth` was invoked during the request.

Perhaps:

1. A `CurrentAuth` object is constructed every time `current_auth` is invoked. It should be constructed once and stored on the request context stack as `current_auth`.
2. CurrentAuth copies the `user` and `auth` attributes on the request stack. Is this still necessary? Why not mutate the CurrentAuth object itself? (Using the existing `add_auth_attribute` helper, as the primary interface is better off read-only.)
3. A new helper function `request_had_auth` can indicate if the CurrentAuth instance exists.
4. Login managers continue to implement `after_request` handlers that do tear down, but only `if request_had_auth()`.

### PR DESCRIPTION
**Pull request by jace**

### CODE REVIEW

File: coaster/auth.py [hunk 1]
```diff
@@ -5,43 +5,114 @@
...
+    if actor and ca.actor is None and value is not None:
+        object.__setattr__(ca, 'actor', value)
+
+
+def add_auth_anchor(anchor):
+    """
+    Helper function for login managers and view handlers to add a new auth anchor.
+    This is a placeholder until anchors are properly specified.
+    """
+    current_auth.anchors._add(anchor)
```
Comment (shreyas-satish) [inline coaster/auth.py]: I'm a little unclear about why `_add` is used here (and in the tests below). Is `AuthAnchors` supposed to be read-only?

Comment (jace) [reply coaster/auth.py]: The request will have an initialisation phrase. Before_request loads the user from cookies, while decorators load anchors (access keys in a URL). These have to be added to current_auth. Once the main view function is underway, current_auth is read-only.

To make this work, current_auth does not publish write interfaces. Writing happens through the helper functions.

**Comment by jace** [comment]
On `actor` instead of `user`:

* In Lastuser, some APIs are for clients rather than users. The client is the authenticated actor.
* In Imgee, an image can be owned by a user, organisation or client (for example, Hasjob employer logos for old jobs posted without a user account). All three are principals, meaning they can own data, but only user and client are actors, meaning they have agency to manipulate data.
* In Hasjob, we have two types of actor: `user` and `anon_user`. Both are first class database-backed objects. Only one of them may exist in a request, and `current_auth.actor` points to whichever.

**Comment by jace** [comment]
I'm uncomfortable with `is_anonymous` and `is_authenticated` being exact opposites of each other. One of them is redundant.

Assuming we continue to develop the anonymous user model (as in Hasjob), where an anonymous user, known only from their browser session, can perform nearly every activity that a logged-in user can, what does `is_authenticated` mean anymore?

Perhaps:

1. `is_anonymous` returns `True` only if `actor.is_anonymous` is `True` (as at present).
2. `is_authenticated` is always `True` when an actor is present. This makes `if current_view.is_authenticated` synonymous with `if current_view.actor is not None`.

If we do this, `is_authenticated` continues to work for apps that don't support anonymous users. Apps that _do_ support anonymous users must take care to examine `current_auth.user` or `current_auth.is_anonymous`, and must particularly use this when showing login buttons.

**Pull request merged**